spec/Section 4 -- Introspection.md

```diff
@@ -291,6 +289,7 @@ Fields
 * `fields`: The set of fields required by this interface.
   * Accepts the argument `includeDeprecated` which defaults to {false}. If
     {true}, deprecated fields are also returned.
+* `interfaces`: The set of interfaces that an interface implements.
 * `possibleTypes` returns the list of types that implement this interface.
```

Comment:
I finally got a chance to start reimplementing this in graphql-js (I'm just starting over, since there are a large number of conflicts and I want to give the entire codebase a scan for relevant changes anyhow.)

This has reminded me of one point I [briefly mentioned earlier](https://github.com/graphql/graphql-spec/pull/373#issuecomment-500554181) but want to call out here.

I intentionally left the `possibleTypes` field definition unchanged in the spec, which means that it's now possible for interfaces to show up in the list (as opposed to only concrete types). While this isn't technically problematic, I suspect that tooling has been written with assumptions that will no longer be true.

For anyone tracking this, I just want to bring attention to this minor "change cost."

Reply:
We should consider alternatives, because I would expect that this breaks clients/tooling.

Reply:
I guess our options are:

1. Add interfaces to `possibleTypes`, which should technically be fine, since it's still a `__Type`, and the spec has never required a specific value for its `.kind`. Filtering for concrete types is as simple as ensuring `.kind` is `__TypeKind.OBJECT`.
2. Continue to only include implementing `__Type` objects where `.kind` is `__TypeKind.OBJECT`
    1. Don't do anything else. Finding relationships between interfaces is still possible, but considerably more challenging, as the tree must be traversed from leaves back to the trunk using the `.interfaces` field.
    2. Add a new field that includes only implementing interfaces.
    3. Add a new field that includes implementing interfaces and concrete types

Option 1 definitely feels the cleanest here... but I'm curious what kind of tools might have made this kind of assumption, and what kind of consequences it may have.

Reply:
@mike-marcacci If you dump entire introspection you need `possibleTypes` only for union and since interfaces you can use `interfaces` field inside object type. An AFAIK it whats everyone currently doing, and here are examples in [graphql-js](https://github.com/graphql/graphql-js/blob/adb5292aea44baa989f3672d6db05f2949fbc0aa/src/utilities/buildClientSchema.js#L247),  [graphql-java](https://github.com/graphql-java/graphql-java/blob/3cbc1adfcf919970cb02fa8eb9a0cbc7fe1c3fab/src/main/java/graphql/introspection/IntrospectionResultToSchema.java#L216) and pretty sure all other implementations.

So I don't think we will actually break any tooling if we add interfaces to `possibleTypes`.

Reply:
Also, in the absolute majority of cases, introspection queries are used by dev tools and breaking them is not such a big deal. Another argument would be that it's opt-in feature meaning it's up to API owners to evaluate if they breaks something or not by starting to use this feature.

Reply:
I agree – this is 100% my stance on this as well. I just wanted to call out the caveat here.